## 4.5 Data Consistency Checks

### Contents

### 1. Exercise 4.5 procedure
### - Importing libraries and files
### - Mixed data type
### - Missing values
### - Duplicates

### 2. Exercise 4.5 task

#### 1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

#### - Importing libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create a variable to access our files and helps us save time reaching them
path = r'C:\Users\boss4048\Documents\10-2024 Instacart Basket Analysis'

In [3]:
# Save the file into df_prods variable, so we can call and use the variable anytime we need
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [4]:
# Test the file we accessed as df_prods variable
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [5]:
# Save the file into df_ords variable, so we can call and use the variable anytime we need
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [6]:
# Test the result from df_ords
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  total_days_since_last_order  
0                        8                          NaN  
1                        7                         15.0  
2                       12                         21.0  
3                        7                         29.0  
4                       15                         28.0  
...                    ...                          ...  
3421078                 18                         29.0  
3421079                 10                         30.0  
3421080                 12                         18.0  
3421081                 12                          7.0  
3421082                 14                         30.0  

[3421083 rows x 7 columns]

In [7]:
# df.describe() function returns descriptive statistics for the numeric values in the dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  total_days_since_last_order  
count        3.421083e+06       3.421083e+06                 3.214874e+06  
mean         2.776219e+00       1.345202e+01                 1.111484e+01  
std          2.046829e+00       4.226088e+00                 9.206737e+00  
min          0.000000e+00       0.000000e+00                 0.000000e+00  
25%          1.000000e+00       1.000000e+01                 4.000000e+00  
50%          3.000000e+00       1.300000e+01                 7.000000e+00  
75%          5.000000e+00       1.600000e+01                 1.500000e+01  
max          6.000000e+00       2.300000e+01                 3.000000e+01

#### - Mixed type data

In [8]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
# Check the result of the column we created
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Use for-loop to find out if there are any mixed data, if there are (Boolean = True), print command will work
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
# Used .astype() function to convert the data type from integer to be string
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
df_test.dtypes

mix    object
dtype: object

#### - Missing values

In [14]:
# Check if there are any missing data in each column from df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# Uses subsetting method to show only the rows that have missing values by applying .isnull() function
# Saved the filtered rows into a new variable (df_nan)
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
# Test the result of nan values dataframe
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
# Check descriptive statistics of df_prods dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [18]:
# Because null values are in product_name column, we will not be able to replace the values with the code below
# df['column with missings'].fillna(mean value, inplace=True) 
# We decided to show only rows that are not missing out into a subset dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
# Check the size of the new dataframe which have non-empty values
df_prods_clean.shape

(49677, 5)

In [20]:
# Another way is to drop the missing values rows by using .dropna() function
df_prods.dropna(subset = ['product_name'], inplace = True)

#### - Duplicates

In [21]:
# .duplicated() function identifies all rows with duplicated values
# Save it into a new variable (df_dups)
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
# Test the result, there are 5 duplicated rows
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
# Check the size of df_prods dataframe before droppping the duplicates
df_prods_clean.shape

(49677, 5)

In [24]:
# Used .drop_duplicates() to drop the duplicated rows
# Save it into a new variable df_prods_clean_no_dups
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
# Test the result and the number of rows dropped to 49,672 rows
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [26]:
# Changed the name of dataframe to look more professional
products_checked = df_prods_clean_no_dups

#### - Tidying Up and Exporting Changes

In [27]:
# Export the clean dataset to our destination folder
products_checked.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

#### 2. Run the df.describe() function on your df_ords dataframe. Interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [28]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  total_days_since_last_order  
count        3.421083e+06       3.421083e+06                 3.214874e+06  
mean         2.776219e+00       1.345202e+01                 1.111484e+01  
std          2.046829e+00       4.226088e+00                 9.206737e+00  
min          0.000000e+00       0.000000e+00                 0.000000e+00  
25%          1.000000e+00       1.000000e+01                 4.000000e+00  
50%          3.000000e+00       1.300000e+01                 7.000000e+00  
75%          5.000000e+00       1.600000e+01                 1.500000e+01  
max          6.000000e+00       2.300000e+01                 3.000000e+01

Answer: Total_days_since_last_order column has min value = 0, which could represent 0 values and NAN values. This applies the same with 'orders_day_of_week' and 'order_hour_of_day' columns

In [29]:
# I used .value_count() function to check if the column has NaN values 
# which represent 0 values on the result of min from using .describe() function
# We can see that this column has both real 0 values and NaN values
df_ords['total_days_since_last_order'].value_counts(dropna = False)

total_days_since_last_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

#### 3. Check for mixed-type data in your df_ords dataframe.

In [30]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Answer: There are no mixed data types in the df_ords Dataframe after executing the code above.

#### 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type. 

Answer: There are no mixed data types in any of the columns of the df_ords DataFrame.

#### 5. Run a check for missing values in your df_ords dataframe

In [31]:
# From using .describe() and .value_counts() functions from prodedures above
# We found out that 'total_days_since_last_order' column contains 0 values and NaN values
# I am investigating why NaN values showed up on this column
# Used subsetting method to filter and display only rows with NaN values
df_nan = df_ords[df_ords['total_days_since_last_order'].isnull() == True]

In [32]:
# Test the dataframe with only NaN values
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  total_days_since_last_order  
0                        8                          NaN  
11                      11                          NaN  
26                      14                          NaN  
39                      11                          NaN  
45                      12                          NaN  
...                    ...                          ...  
3420930                 12                          NaN  
3420934                 18                          NaN  
3421002                 19                          NaN  
3421019                 15                          NaN  
3421069                 11                          NaN  

[206209 rows x 7 columns]

There are total 206,209 rows with NaN values.

#### 6. Address the missing values using an appropriate method

I am suspicious that 'order_number' column at order_number = 1 might cause this trouble. My hypothesis is 'order_number = 1 might contain only NaN values. To prove it, I will check how many rows that order_number = 1 has, if it is equal to 206,209 rows (match the rows of NaN values). That means all rows when order_nyumber = 1 contain all NaN values.

In [33]:
# Used subsetting method to filter and show only order_number = 1
df_order_number_1 = df_ords[df_ords['order_number'] == 1]

In [34]:
# Test the result
df_order_number_1

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  total_days_since_last_order  
0                        8                          NaN  
11                      11                          NaN  
26                      14                          NaN  
39                      11                          NaN  
45                      12                          NaN  
...                    ...                          ...  
3420930                 12                          NaN  
3420934                 18                          NaN  
3421002                 19                          NaN  
3421019                 15                          NaN  
3421069                 11                          NaN  

[206209 rows x 7 columns]

Now we see that all rows order_number = 1 contain all NaN values (206,206 rows).
It makes sense that order_number = 1 contains empty (NaN) values on total_days_since_last_order (days), 
because those are new customers, they have never ordered from the company before.

#### 7. Run a check for duplicate values in your df_ords data

In [35]:
# Used .duplicated() function on the cleaned dataset (df_ords)
df_dups = df_ords[df_ords.duplicated()]

In [36]:
# Test the result
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, total_days_since_last_order]
Index: []

Answer: No duplicates found

#### 8. Address the duplicates using an appropriate method.

Answer: There are no duplicates found because from the code avove, it returned empty. However I would use this code below to delete duplicated rows from the dataset.
df_cleaned_data = df_ords.drop_duplicates()

In [37]:
# Changed to be a succint name
orders_checked = df_ords

#### 9. Exportm your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [38]:
# Used .tocsv() function to export the cleaned files to our destination folder
products_checked.to_csv(os.path.join(path, '02 Data', 'Prepared data', 'products_checked.csv'))
orders_checked.to_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_checked.csv'))